src/Workspaces/Core/Portable/Workspace/Host/PersistentStorage/IChecksummedPersistentStorage.cs

```diff
@@ -5,6 +5,64 @@
 using System.IO;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PersistentStorage;
+
+namespace Microsoft.CodeAnalysis.PersistentStorage
+{
+    internal readonly struct SolutionKey
+    {
+        public readonly SolutionId Id;
+        public readonly string FilePath;
+        public readonly bool IsPrimaryBranch;
+
+        public SolutionKey(SolutionId id, string filePath, bool isPrimaryBranch)
+        {
+            Id = id;
+            FilePath = filePath;
+            IsPrimaryBranch = isPrimaryBranch;
+        }
+
+        public static explicit operator SolutionKey(Solution solution)
+            => new SolutionKey(solution.Id, solution.FilePath, solution.BranchId == solution.Workspace.PrimaryBranchId);
+    }
+
+    internal readonly struct ProjectKey
+    {
+        public readonly ProjectId Id;
+        public readonly string FilePath;
+        public readonly string Name;
+
+        public ProjectKey(ProjectId id, string filePath, string name)
+        {
+            Id = id;
+            FilePath = filePath;
+            Name = name;
+        }
+
+        public static explicit operator ProjectKey(Project project)
+            => new ProjectKey(project.Id, project.FilePath, project.Name);
+    }
+
+    internal readonly struct DocumentKey
+    {
+        public readonly ProjectKey Project;
+
+        public readonly DocumentId Id;
+        public readonly string FilePath;
+        public readonly string Name;
```

Comment:
so, it was never clear to me from our workspace API if we guarantee we always have one or the other.  So this is resilient to a case where a Project has Documents with names, but without filepaths.  As long as 'name' is unique within the Project, it will properly work.